### PR TITLE
[OSD-17447] Gather hyperinfo from all associated AWS accounts

### PR DESCRIPTION
**Pull request by bergmannf**
This is a PR that will attempt to fetch resources from all AWS accounts associated with a hypershift accounts.

This is mostly, because it is really annoying to log into all those accounts one-by-one and finding if everything is there, so this should gather at least of the resources and print them in table form (and graphviz dot notation, as the tables are quite big).

Still in draft to get some feedback if this is considered useful.

### CODE REVIEW

File: cmd/hypershift/cmd.go
Comment (clcollins) [inline cmd/hypershift/cmd.go:8]: I think this comment is a copy/paste remnant :D

File: cmd/hypershift/cmd.go
Comment (clcollins) [inline cmd/hypershift/cmd.go:9]: Is this intended to be `osdctl hypershift` for the command?  

If so, I think this would be better as a noun-verb subcommand for `cluster`, command, given the thing you're going for here.  Maybe something like `osdctl cluster get-cloud-resources` or something along those lines.

If it can be generic-ish, like this, someone else can do something similar for non-hypershift clusters.  I think there's benefits for both.  Obv not needing to gather management cluster info, etc, but getting the resources themselves would be an excellent advantage for both types!

File: cmd/hypershift/info.go
Comment (clcollins) [inline cmd/hypershift/info.go:162]: Is there a benefit to using the channels to collect errors?  Not opposed, just curious on the thinking, for my own education.

File: cmd/hypershift/info.go
Comment (clcollins) [inline cmd/hypershift/info.go:142]: Would you be opposed to wrapping informational messages like these in a check for the verbose global flag? By default less is more, especially to allow folks to pipe output to other things.

File: cmd/hypershift/info.go
Comment (clcollins) [inline cmd/hypershift/info.go:152]: Is there a reason for using the `log` package here instead of `fmt`?  Most of the rest of osdctl isn't using log output, and while I'm not opposed, I'm not sure what the benefit of timestamps on a manually run too is.  It's valuable for actual log files if you're looking back at the history, but I think it just adds noise to the output for a CLI tool.

EDIT: Nevermind, looks like we're using log in a lot of places!  I take back my comment.

File: pkg/provider/aws/client.go
Comment (clcollins) [inline pkg/provider/aws/client.go:104]: Just a note that there's an effort that's almost completed to port the AWS stuff to the v2 SDK: https://github.com/openshift/osdctl/pull/411

It might be worth updating this PR to that before merging.

**Review by clcollins — CHANGES_REQUESTED** [review]
I love this PR - being able to visualize the resources for a cluster will be a great addition, especially for troubleshooting!

Just a couple small requests or requests for info for my own edification.

File: cmd/hypershift/cmd.go
Comment (bergmannf) [reply cmd/hypershift/cmd.go:9]: Yes I totally agree, I was also thinking that having a similar command (or this command with the right flags) provide similar output for both - given that one of the cluster's that should be collected is the customer cluster anyhow that whole part should be 90% reusable (maybe except the VCP endpoints).

File: cmd/hypershift/info.go
Comment (bergmannf) [reply cmd/hypershift/info.go]: Yea I guess I just looked at one of the commands that does, but I agree with the 'timestamps are not that useful' sentiment in this case - so I'm fine switching to just fmt printing, so we don't add more log code if we don't really want it.

File: cmd/hypershift/info.go
Comment (bergmannf) [reply cmd/hypershift/info.go:168]: I think it's one of two ways (at least that I know of), to indicate an error during execution of the goroutine.
As the goroutine doesn't return a value at all, there must be another way to check for an error to have occurred.

I think the alternative would be use a single channel to also gather the output and use an aggregated type that wraps the expected return value & an error, so once the value is retrieved the caller can check if an error occurred:

```
type retrievable interface {
	clusterInfo | managementClusterInfo | privatelinkInfo
}

type ChanReturn[T retrievable] struct {
	Value T
	Error error
}
```

Then just use a `chan[ChanReturn]` and pass that to all async functions.

If there is a more synchronous-looking/behaving way I'd be happy to use that - the current mix between output-argument passed in and channel is quite ugly.

Comment (clcollins) [reply cmd/hypershift/info.go]: Thank for the explanation!  This sounds good to me, and I didn't have any issues with it - just trying to learn more myself, thanks!

**Review by mjlshen — APPROVED** [review]
/approve
/label tide/merge-method-squash

**Comment by AlexVulaj** [comment]
/hold

Want to make sure this gets merged after the aws-sdk-v2 migration that should be merged very soon

File: cmd/cluster/graphviz.go
Comment (AlexVulaj) [inline cmd/cluster/graphviz.go:1]: Can we move this file out of `cmd` since it doesn't hold a command struct? I also think it could have more general usage in the future. I'm thinking somewhere in `pkg`

File: cmd/cluster/hypershift_info.go
Comment (AlexVulaj) [inline cmd/cluster/hypershift_info.go:41]: Could we shorten this to `cobra.ExactArgs(1)` ?

File: cmd/cluster/graphviz.go
Comment (bergmannf) [reply cmd/cluster/graphviz.go:1]: I moved the rendering part out.

Generating the 'graph' structure is now in the hypershift_info.go file, because otherwise there would be a circular dependency - so every code that wants to render graphviz has to generate the graph and just pass it to the render function.

Comment (AlexVulaj) [reply cmd/cluster/graphviz.go:1]: Love that approach!

**Comment by AlexVulaj** [comment]
/lgtm

confirmed my last question with @bergmannf here: https://redhat-internal.slack.com/archives/CNJPJSG3Z/p1692189668859019

**Pull request merged**
